src/link/Elf.zig

```diff
@@ -1662,11 +1685,14 @@ pub fn resolveSymbols(self: *Elf) !void {
             _ = self.objects.orderedRemove(i);
         } else i += 1;
     }
+    // TODO This loop has 2 major flaws:
+    // 1. It is O(N^2) which is never allowed in the codebase.
+    // 2. It mutates shared_objects, which is a non-starter for incremental compilation.
```

Comment:
The reason we have this as well as a similar loop for objects (well, relocatables) is that we parse *all* relocatables from all archives and *all* shared objects that appear on the linker line. This means we end up with an upper bound on all potentially required inputs. By this I mean that not all relocatables from archives or shared objects will necessary be needed and actually may yield duplicate symbols errors or lead to an outright broken output image. Therefore, we do symbol resolution twice: first with all objects parsed and extracted, then we mark any unreferenced object as dead, prune the sets, and re-run the resolution again. This non-standard algorithm matches what `mold` does, however it is not what `lld` does. `lld` for instance has the concept of *lazy* inputs which are analysed only when required. My understanding after studying the work of Rui/`mold` is that the `lld` approach which is effectively the legacy approach is slower than `mold`'s if inputs are parsed in parallel.

Reply:
As for the O(N^2) problem, it can be addressed without any architectural changes, it just needs to not call orderedRemove inside the loop. You can implement this kind of thing in O(N) by doing one loop only and keeping track of the number of removals as you go along.

Regarding garbage collection, subsequent updates might cause a different set of things to remain alive, so the data must not be mutated. Instead, just like the rest of the compiler pipeline, it should create read-only transformations of state from one thing to another, selectively invalidating and redoing parts of the pipeline depending on what changed since last time.

As a reminder, I want to focus on single-threaded linking that happens concurrently with the rest of the zig compiler first, before exploring multi-threaded linking. You did this in the other order for MachO but I think it will yield a better outcome to do instead as I suggested for ELF. I understand the urge to compare directly the performance against mold or lld but that is completely missing the point. We need to lean into our unique advantage in the Zig project which is the fact that the compiler and linker are tightly coupled and we can therefore start linking and compiling at the same time. In this case a single-threaded linker may even yield the best performance since it has less overhead, as long as the compiler also has work to do in the meantime.

Reply:
Yeah, I like that plan, and looking forward to the outcome!